contributors/design-proposals/sig-cli/apply_v2.md

```diff
@@ -0,0 +1,94 @@
+# Apply v2
+
+## Background
+
+`kubectl apply` reads a file or set of files, and updates the cluster state based off the file contents.
+It does a couple things:
+
+1. Create / Update / (Delete) the resource based on the resources
+2. Only set / delete fields defined in the file either currently or in during last update from the file
+  - Retain updates to fields made by controllers that were never managed in the file
+3. Merge fields based on the field name, merge lists based on a field contained within the list elements
+
+The apply logic has been difficult to maintain from a contributor perspective, and difficult
+to understand from a user perspective.  Much of this pain has comes from the `apply` command out
+growing its implementation.  In order to continue to support apply and provide better user visibility
+into what it is doing, the command must be re-architected to break client dependencies on the server-side
+implementation, address gaps in diff-patch semantics preventing correctly applying certain fields,
+and to organize the code in a modular and extendable manner.
+
+Reducing the code complexity and improving the correctness will also make it simpler to move this logic
+to the server as part of an API in the future, while also addressing the immediate needs of the client.
+
+## Changes
+
+### Use read-update instead of patch
+
+#### Why
+
+Issues with updating the object by calculating and sending a patch to the server
+
+- Patch API is not fully featured enough to represent all of the operations required
+  to update an object and frequently has to be expanded with new side-directives describing
+  how the server should apply the patch.  Updates are done in the form of adding new
+  merge directives to the patch format on the server, and invoking them from the client.
+- Having to update both client and server to fix bugs introduces challenges to supporting
+  version skew, as the client will depend on the server supporting a patch semantic.  There
+  is currently no versioning of the patch semantics, so the client cannot reliably know what
+  patch semantics the server supports.
+- Patch syntax is not human readable, making it hard for humans to evaluate the end state
+  that will result from the patch.  `--dry-run` is nearly useless for understanding the
+  end state that will result from apply, and implementing a diff operation has been exceedingly
+  copmlex.
+- Generating a patch to make a change is more complicated than simply making the change to
+  the object read from the server.  The apply command reads the live object from the server
+  and diffs it against the local configuration.  Building a patch containing special merge
+  directives is a more complicated approach than modifying the object read from the server
```

Comment:
True.

However:
* My impression is that most of the complexity is not in the directives but in the API schema and defaulting behaviors, as I mentioned in the [server-side apply proposal](https://docs.google.com/document/d/1G4L4Kdyfg80DDuVOXHY0UbFuqDO29kECnOacVlseiFw/edit#).
* The schema needs to be accurately communicated to the client from the server (e.g., via extended OpenAPI) for client-side patch operations in order to support API aggregation and CRD, which re-introduces versioning and skew issues until the syntax stabilizes, but actually would improve skew behavior once it does.
* We need to support these patch behaviors independently of apply.

Reply:
> My impression is that most of the complexity is not in the directives but in the API schema and defaulting behaviors, as I mentioned in the server-side apply proposal.

Constructing a set of directives to order and delete items from a list results in more complicated code than just deleting the objects from the list.

> The schema needs to be accurately communicated to the client from the server (e.g., via extended OpenAPI) for client-side patch operations in order to support API aggregation and CRD, which re-introduces versioning and skew issues until the syntax stabilizes, but actually would improve skew behavior once it does.

Agreed.  This proposal should also work independently of whether it is done on the client or server.  The focus of this proposal is on rewriting the 3-way diff code to be maintainable.  The inputs will be the 3 versions of the object and the output will be the new object.  

> We need to support these patch behaviors independently of apply.

Yes, but why do we need to use them in apply?